### ISSUE TITLE
Missing shared dependencies across dynamic imports

### ISSUE THREAD

**Issue by rixo** [PepsRyuu/nollup#85]
Bear with me...

If I have 2 dynamic imports in my entry points, and both of them import another module, that itself imports another module... And if only the 2nd dynamic import is actually executed, then the transitive dep module is missing, which results in the following error:

~~~
Uncaught (in promise) TypeError: modules[number] is not a function
~~~

I've pushed a reproduction case on the [bug-shared-dynamic-imports](https://github.com/rixo/nollup-test-case-output-dir/tree/bug-shared-dynamic-imports) branch of my playground repo of the moment.

To illustrate, here what a simple `main.js` would look like:

~~~js
if (false) import('./dyn1.js')
import('./dyn2.js')
~~~

Where `dyn1.js` and `dyn2.js` would both have this code:

~~~js
import dep from './dep.js'
console.log(dep)
~~~

And `dep.js` itself would be:

~~~js
import sub from './sub.js'
export default sub
~~~

It's this `sub.js` module that would be missing from the `dyn2.js` "chunk". It would be present in `dyn1.js`, but since it is skipped at runtime... Still missing.

**Comment by PepsRyuu** [comment]
Thanks for the report! The ```file.check``` exists to prevent a circular dependency. So will need to investigate how this affects that.

**Comment by PepsRyuu** [comment]
Digging into this, I think it's best for the entire chunking code to be refactored. I think I can make it a lot cleaner and more reliable now that the different scenarios have been identified (missing deps, naming collisions). 

I think I'll focus on wrapping up some of the smaller bugs first so that they're not being held up, and then do a large refactoring of the chunking code which will cover a lot of different files.

**Comment by rixo** [comment]
Makes sense, if you can think of a better global organisation.

BTW I open PRs mainly for informative purposes. It's often clearer just to show some code. Don't be afraid of closing or implementing better ways, I have no strong attachment to them.

If we're in for a refactor, do you think we can come up with a way to not include shared dependencies with every chunk? I'm afraid it might not scale too nicely with the highly code splitted apps we're trying to do these days (just a hunch).

More importantly, it seems that Nollup rebuild time increases linearly with project size / complexity. My understanding of what Nollup is doing is that we could achieve almost constant time on rebuild, essentially reprocessing only the file(s) that have changed. Constant rebuild time is the Graal here, it would put Nollup on par with the trending no-bundle solutions. I'm gonna throw some timers in there and try to make a better idea of what could be done, but am I missing something obvious? Any thoughts on this?

And since we've started the crazy talk... Just floating an idea but what would you think of replacing the homemade module loader with SystemJS? It would bring perfect ESM emulation (including live bindings), which is in the air of the times -- ES modules. Rock solid, and externally maintained. It also supports Node, so we could do with a single module system instead of conditions throughout the codebase... I know this would be a dramatic change, but I'm curious about how you feel about this idea.

**Comment by PepsRyuu** [comment]
No worries at all!

Shared dependencies is an interesting one. It's a known issue, but I didn't see it originally as a scaling problem. My only concern here is what performance impact if any would it have to figure out the best chunking combinations. That's more of a production task, so it would be good to figure out what a good middle ground for development is that isn't as processing intensive.

That's another reason why I want to do some refactoring. I fear with all of these quick bug fixes, it's gotten a bit messy and doing a lot of unnecessary work. Also at the moment, Nollup goes through the entire dependency tree on a rebuild, and checks if it has that file in the cache or has to be invalidated. If invalidated, it goes through the plugin lifecycle for that module. This can definitely be optimised. Also pretty sure I can throw in a bunch of other optimisations like caching the bundle strings which I'm not doing currently. Remember that Nollup is a couple of years old now, so the knowledge I have now versus then can definitely improve a lot of things!

I love crazy talks. :D But SystemJS isn't really a good option for a few reasons: Lack of compatibility with ```module``` and ```require```, and inability to improve performance. For example, live binding isn't implemented not because of a technical constraint, but because it's a huge performance drain and breaks debugging symbols. So perfect ESM emulation isn't actually desirable. (I may consider live-bindings to be an option though that can be enabled separately at some point).

**Comment by rixo** [comment]
Exciting!

I have no idea of the perf impact of duplicating the shared dependencies, but my concern is that I get whole libraries bundled with every dynamic chunk (e.g. `import mylib from 'mylib'`, also Svelte runtime -- do people get a whole React in each of their dynamic chunk?). In a typical code splitting app, every page is a dynamic chunk, so it seems to waste a lot of parsing...

Maybe instead of figuring out what shared dependencies should be chunked together, all modules could be lazy loaded from the client? As far as I can tell, this is a problem we have only with dynamic imports, this means it is possible to load the modules we need just in time before resolving the dynamic import promise.

That being said, I haven't tested thoroughly but my first impression is that it doesn't seem to have a huge impact.

I looked a little closer into rebuild time though... A sizable chunk of time is passed in the `generate` function so, indeed, caching the generated code could save hundreds of ms from rebuild time, especially when project is growing. I would take this.

Appart from that, the increased rebuild times I have seen so far were mostly caused by my barbarian fix it seems. Hu hu. It bypasses the parse "cache". So, if rebuild time does increase linearly with project size, it probably does so several order of magnitude slower that I initially thought. And so, in practice, it is probably negligible.

I'd love to do more benchmarking with big code splitting projects, but the dynamic import bug (current issue) prevents me. When you've got some possible fix, I'd be happy to try it if you want.

**Comment by PepsRyuu** [comment]
Just a small update on this, currently in the process of refactoring the entire code. Will be a bit of time still before it's ready but just saying that this is being addressed. 🙂

**Comment by rixo** [comment]
That's great news! Thanks for the update.

I managed to patch together a somewhat working version for one of my use case, and it is very promising. I can't wait for the well crafted version!

I've got 3 Svelte projects awaiting Nollup-ization -- including Sapper, the official Svelte app framework, and Routify, the rising star :smile: Well, I absolutely don't mean to pressure you, just sharing my enthusiasm... I can wait actually, I have plenty of other things to do :sweat_smile: 

But ready to try / review anything if you need!

**Comment by PepsRyuu** [comment]
There's a work in progress on https://github.com/PepsRyuu/nollup/tree/Refactor if you want to check it out. It's mostly just missing ```emitChunk``` at the moment. Still need to do some rigorous testing including for this issue.

**Comment by rixo** [comment]
Great! Trying this in the complicated setups like Sapper to see if I can make it work :D

**Comment by PepsRyuu** [comment]
Branch updated again with ```emitChunk``` working. Still needs more testing but as far as I can tell the situation regarding missing dependencies in dynamic imports and name conflicts should be resolved. The hooks in general are more accurate. Would be great to get your feedback on it and see if it addresses the issues you've been having.

As for next steps: 

* More test cases and tidying those up in general.
* More refactoring, some stuff can still be done better I believe.
* I want to investigate performance more.

**Comment by rixo** [comment]
I have managed to make it work with a pretty complex project I'm currently working on. So far so good regarding the dynamic imports.

Now I need to do some work on the actual app, to see if I don't have some edge cases, but what I've seen so far is encouraging. Apparently the the freeze / race I was encountering with my patched montage version might be gone.

In the process of hooking into my app, I had to play with the public dir things. I'm having second thoughts about it... About the discarding some files part at least. I'll add more details about my findings in #84.

**Comment by rixo** [comment]
I've got a funky one. HMR crashes when renaming a file and simultaneously changing the file that imports it :sweat_smile: 

It happens on Refactor but not on master.

I've prepared some [detailed explanations](https://github.com/rixo/nollup-test-case-output-dir/tree/bug-rename) on the bug-rename branch of a test repo I had handy!

**Comment by PepsRyuu** [comment]
I can't trigger the static bug. This would lead me to suspect it's a timing issue based on the operating system?

![image](https://user-images.githubusercontent.com/6637882/83505838-45340500-a4be-11ea-8744-c904013d4d65.png)

Perhaps it can be replicated with a test case using the stubbed file system. Dynamic one fails for me at the moment because of a file name encoding issue because of Windows paths and additional quotes in the key. That can be fixed easily.

**Comment by rixo** [comment]
Yes, I can confirm there is a timing aspect to the static condition. I do get either the reported error, either success.

By changing the order and unparallelizing operations to rename first, the problem seems more systematic. I haven't had a success with this (yet? you never know with races...).

I've pushed the changed order to the test repo, if you want to try.

**Comment by rixo** [comment]
`import.meta`? What does Nollup do with `import.meta`?

**Comment by PepsRyuu** [comment]
Same as Rollup, for ```import.meta.ROLLUP_FILE_URL_referenceId```, and any custom use case with ```resolveImportMeta``` hook. There's no custom ```import.meta``` properties or anything. The problem before the refactor was that the ```import.meta``` hooks were triggering before they were supposed to, and this can have an impact when emitting chunks. The refactor corrects this, but like dynamic imports, it uses an object map so that it allows for caching of files for faster builds.

**Comment by PepsRyuu** [comment]
Still unable to trigger the static version with the latest commit. :(

**Comment by rixo** [comment]
ok... I've done some debugging of the static condition. I've managed to catch the __hmr changes message both when the problem happens, and when it doesn't.

It is failing when the change to `main.js` (0) is first in the changes array:

```
changes: [{id: 0,…}, {id: 2,…}]
```

It works when the change to the newly created module (2) (e.g. foot, when renaming foo -> foot) is first in the changes array:

```
changes: [{id: 2,…}, {id: 0,…}]
```

However, even with this knowledge, I don't manage to trigger the bug consistently in my test case (btw, just realized my `serial` implem in `bug.js` was broken -- pushed a fix ^^).

So... What can influence the order of the changes array? Is this deterministic? Does it depends on the order of the FS watch events (but I don't manage to control the bug by playing with the fs change order)?

**Comment by PepsRyuu** [comment]
I updated the branch fixing the dynamic issue.

I managed to reproduce the problem with these steps:

* Start Nollup, load the page.
* Running the bug script, no error.
* Keeping Nollup running, refresh the page.
* Run the bug script, throws an error.

It's a mismatch of state between the server and the client. Current delta is checked by comparing the length of the file cache with the previously length. This is invalid, because not all files in the file cache are in a bundle. They were in the previous version of Nollup.

**On load:**
```
Files: 0, 1
Bundle: 0,1
Files After Script: 0,1,2
Delta After Script: 2 (since it's new) and 0 which was modified
```
**On Refresh:**
```
Files: 0,1,2
Bundle: 0,2
Files After Script: 0.1,2
Delta After Script: 0 (it imports 1, but that doesn't exist in bundle, so it crashes)
```
Should be able to come up with a fix.

**Comment by PepsRyuu** [comment]
Pushed what I hope is a fix for this! :D

**Comment by rixo** [comment]
> Pushed what I hope is a fix for this! :D

Oops. I was going to post the following... I don't know if its still current, but I'm posting it FYI. I'm going to test the new branch now.

---

The new branch fixes the crash with dynamic imports but apparently it fails to reload the imported (and newly recreated) module.

I now get this log:

~~~
> foot
foot
> foo
foo
foot
foo
foot
foo
~~~

Expected:

~~~
> foot
foot
> foo
foo
> foot
foot
> foo
foo
> foot
foot
> foo
foo
~~~

The `>` is from the `console.log` inside the imported module (`foo` or `foot`). It should run every time, since the file has been deleted and recreated, it should be considered new, right? I've pushed a change to my bug script to make the exported value unique, to better show that the new content of the file is not reflected.

Aaah, yes indeed for static, I managed to not reproduce it too now, if I restart Nollup before each run! So that must be the same thing. So the order in `changes` is only relevant when the delta in incomplete?

Here again, I'm a bit surprised by the log I get, though:

~~~
> foot
foot0.41476973700899555
> foo
foo0.6385670261616962
foot0.41476973700899555
> foot
foot0.18909642331451715
~~~

I would have expected:

~~~diff
 > foot
 foot0.41476973700899555
 > foo
 foo0.6385670261616962
- foot0.41476973700899555
 > foot
 foot0.18909642331451715
~~~

The first run is OK, the second one too (first time Nollup sees the `foo` file), but on third... 

Apparently, the update to `foot` is processed first, before the one to main. Since `foot` has no accept handler, the update bubbles to `main`, that is reloaded, and so we get an extra console.log from `main`.

I think the fact that the imported module is processed before its importer might be defensible but it's suboptimal, because we're re-running a stale module (`main`). In `rollup-plugin-hot`, resolving accept handlers and actually reloading modules are 2 separate passes. During the resolving pass, each time an update bubbles to a module (that is, from its non accepted importer), then its pushed to the end of the reload queue. This way, importers are reloaded after imported modules. This is a very crude way of sorting dependencies, and it can break with more modules, but I believe it's what protects `rollup-plugin-hot` from the bug in the test case with only 3 changes / 2 files in each updates.

That being said, even if we accept that deps are not sorted in the `changes` array, then the log line I've flagged is not correct. At this point, if `main` is reloaded before `foot`, then main should still import `foo.js` (and either we should crash if the deleting of `foo` has already been processed, or we should log the stale `foo` value).

This seems to be related to the other problem, with dynamic imports.

**Comment by rixo** [comment]
With the last push, apparently no more crashes in either condition, with or without restarting Nollup.

But I still get the same behaviour as just described about what I think is stale content.

We're making progress! Hold on! ^^

**Comment by PepsRyuu** [comment]
Updated branch again. In a nutshell, dynamic dependencies were not being tracked as dependencies at all, so HMR didn't detect them properly and bubble those dependency changes to the files that imported them. The bug repro code works for me now.

I don't think I can do anything about accept handlers running for each module in the changes array. It's easy in the case of a single accept handler, but when you start looking at multiple accept handlers (both in one branch or multiple branches), it gets incredibly messy. I think the objective at the moment should be ensuring that after all accept handlers have resolved, that the state is consistent with a full reload.

**Comment by rixo** [comment]
It was more about the order in which they are run than not running some of them. But yeah, I agree, let's make it consistent first. And maybe also confirm that it really has a damaging effect in real world.

Just tested the new version in the test case. Seems to fix everything :tada: No more stale log either apparently.

I'm merging this into my patched branch and resume using it in my project to see how it fares in real world. I'm precisely working on adding and moving lots of files around just like in the test case.

**Comment by rixo** [comment]
Update: still going strong. I haven't detected any other glitch with this branch, in extended usage :)

I had a false alarm, though. Turned out I had an infinite loop in my code, with some dynamic imports peppered throughout. With SystemJS (in rollup-plugin-hot), the UI had some time to breath on the imports, but not with Nollup... I almost reported that dynamic import should always be async in Nollup, but apparently they are (right?). So ok, just too fast, not a bug.

Ok! So at this point, I'm certainly always testing the same workflows in a circle with my app... I'm now gonna update my Sapper experiment. There were people impatient to try it there, and they rapidly hit the same issues as me... This should gather us a larger scope of feedback!

**Comment by PepsRyuu** [comment]
Dynamic imports are always async, even if they've been loaded previously. It'll of course be faster on a reload because there's no file to download, but it still ensures that a promise is returned so that the code that imports the module doesn't break.

Looking forward to hearing the feedback! If you need any assistance on the integration let me know.

**Comment by rixo** [comment]
Yes, actually I do need some help! I think this should rather go in #84, but I have a lot of other information to share there, I need to sort it all out :sweat_smile:

In short, here's my problem with Sapper... I'm using Nollup's middleware directly (mainly because I need to emulate `rollup.watch(...)`). I need the bundle and dynamic chunks to be served from URL `/client/*`. I also need the `/client` prefix **not** to appear in the bundle's entries' `fileName`, because Sapper uses them to generate the manifest (and so the browser would try to load `/client/client/main.js`, for example). That means `virtualWrite` as we've discussed it doesn't work.

I can get close by mounting the Nollup middleware on `/client`:

~~~js
app.use('/client', nollupDevMiddleware)
~~~

But then Nollup doesn't know about it and tries to load dynamic imports from, for example, `/about.[hash].js` instead of `/client/about.[hash].js`. I tried to hack Nollup's `__imports__map` (approximative name) but it failed further down the road...

Any idea how I could get `about.[hash].js` as `fileName` in the bundle entry, but to have the files, including dynamic imports, served at `/client/about.[hash].js`? I'm just after a monkey patch for now...

**Comment by rixo** [comment]
I have still met the error (`TypeError: modules[number] is not a function`, if I remember correctly) with my monkey patched version.

It happened in a pretty esoteric situation, though, involving multiple renames of the same file(s), including back to its initial name. One time it also threw the server into outta space, and I had to restart it. I haven't really diagnosed the problem.

Maybe it's caused by my patches. They do that sometimes... Anyway, I don't think it's a blocker to carry on. Will see if it pops again.

**Comment by PepsRyuu** [comment]
Released in ```0.12.0```
